design/002-secret-generator-group.md

```diff
@@ -0,0 +1,277 @@
+This custom resource contains all the configuration details to create a secure value.
+It can be referenced from a `Kind=ExternalSecret` or `Kind=PushSecret`. **or embedded** into it.
+
+The controller generates the value and passes it onto the `Kind=Secret` with all the values
+that have been fetched from the provider.
+When used with a `Kind=PushSecret` it will generate the value and push it to the configured provider.
+
+A generator generates a known set of outputs, e.g. `username` and `password`, `private_key` and `certificate` or others.
+
+#### Example CRD
+
+The following custom resource `Kind=VaultPKIBackend` contains all the parameters for issuing a certificate via HashiCorp Vault PKI engine.
+
+```yaml
+apiVersion: generator.external-secrets.io/v1alpha1
+kind: VaultPKIBackend
+metadata:
+  name: my-vault-pki-creds
+spec:
+  # specific params for generating
+  commonName: example.com
+  altNames:
+  - example.org
+  - localhost
+  # ... more opts
+  auth: {} # service account etc. etc.
+```
+
+This resource can be **embedded or referenced** from an ExternalSecret or PushSecret.
+The ExternalSecret controller generates the secret value and it can be post-processed with
+existing features like template or key-rewrite.
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
+  dataFrom:
+
+    # embedded
+    - generator:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: VaultPKIBackend
+        spec:
+          # same as above
+          commonName: example.com
+          altNames:
+          - example.org
+          - localhost
+          # ... more opts
+          auth: {} # service account etc. etc.issuerRef: "default"
+
+    # the generator also can be referenced like here
+    - generatorRef:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: VaultPKIBackend
+        name: foobar
+```
+
+
+<details>
+<summary>PushSecret example</summary>
+
+```yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: PushSecret
+metadata:
+  name: "hello-world"
+spec:
+  secretStoreRefs: [] # ...
+  selector:
+    # we can consider supporting multiple generators per PushSecret
+    # but need to take care of key collision
+    generatorRef:
+      apiVersion: generator.external-secrets.io/v1alpha1
+      kind: VaultPKIBackend
+      name: foobar
+  data: {} # ...
+```
+</details>
+
+<details>
+<summary>Password Generator Example</summary>
+
+```yaml
+apiVersion: generator.external-secrets.io/v1alpha1
+kind: Password
+metadata:
+  name: foo
+spec:
+  characterSet: "..." # TBD
+  length: 42
+---
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
+  dataFrom:
+    - generatorRef:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: Password
+        name: foo
+```
```

Comment:
I know it is an example , but I just need to ask! This example would be exactly the same as generating a Kubernetes Secret with random password, right? In this case, the lifecycle (thinking about TTL here) of the generated password would be in the generator CRD or would it be up to the External Secrets to handle that?

What would happen if ESO pod went down (as I am assuming we really have no place to save the generated password other than in memory)? 😄

Reply:
It would be up to the ExternalSecret. A user would have to choose a `refreshInterval` that suites the rotation requirements. That's how you would rotate passwords. With a `refreshInterval=0` it will be created once. 

> What would happen if ESO pod went down (as I am assuming we really have no place to save the generated password other than in memory)? 

The password/credential created from a generator is considered ephemeral. 
With an `ExternalSecret`: the password is generated in-memory and stored in the `Kind=Secret` it manages. 
With an `PushSecret` it would be generated in-memory and pushed immediately to the external provider.

When ESO goes down there is no issue with that. The stored password stays where it is. ESO will eventually reconcile and regenerate the secret if desired.

Reply:
> When ESO goes down there is no issue with that. The stored password stays where it is. ESO will eventually reconcile and regenerate the secret if desired.

If we have an ExternalSecret with `RefreshInterval=0` and ESO goes down-and-up, the password will be recreated where we wanted it to not recreate. for longer `RefreshIntervals` (e.g. 720h or more), this can also be an issue. I guess what is troubling me is that I don't see how we are controlling a "Lease logic" in the design (i.e. an extra information in ExternalSecret status, or an annotation in the generated Secret?)

Reply:
> If we have an ExternalSecret with RefreshInterval=0 and ESO goes down-and-up, the password will be recreated

The ES should be reconciled once: Refresh is skipped if it has been reconciled: https://github.com/external-secrets/external-secrets/blob/main/pkg/controllers/externalsecret/externalsecret_controller.go#L448

> I guess what is troubling me is that I don't see how we are controlling a "Lease logic" in the design

I wouldn't try to be smart here, instead i would hand over that responsibility to the user. The user knows the lease (it's usually not dynamic) and refresh can be controlled with the refreshInterval (which should be lower than the lease time anyway). Thinking about it: even if we know the expiry time of a secret we still would need to make assumptions about when we should reconcile: 1h before expiry? 10s?
We can not know that and we would either want to add _another_ field to the crd to give the user control of that behaviour or just use the refreshInterval and then there's no need for a LeaseID. :thinking:  :shrug: WDYT?